src/nvim/lua/vim.lua

```diff
@@ -55,29 +55,30 @@ end
 
 function vim._load_package(name)
   local basename = name:gsub('%.', '/')
-  local paths = {"lua/"..basename..".lua", "lua/"..basename.."/init.lua"}
-  for _,path in ipairs(paths) do
-    local found = vim.api.nvim_get_runtime_file(path, false)
-    if #found > 0 then
-      local f, err = loadfile(found[1])
-      return f or error(err)
-    end
+  local paths = "lua/"..basename..".lua lua/"..basename.."/init.lua"
```

Comment:
Is this also working for packages? ` ` as a separator seems extremely fragile (and was an implementation detail that wasn't meant to leak out), and I don't think `do_in_path_and_pp` is handling it for the `pack/*/start/*/%s` branch (which btw is code we wanna change anyway, doing a shell glob over packages and rtp dirs in every call is too slow)

Reply:
The only code path from `require()` besides Lua's own search paths is through `vim._load_package -> vim.api.nvim_get_runtime_file -> do_in_runtimepath -> do_in_path_and_pp -> do_in_path`. The only calls to the callback are in `do_in_path` here: https://github.com/neovim/neovim/blob/b35de6c525e7c8ad9d64d42f57f348dff608b9e5/src/nvim/runtime.c#L102-L137
If `path` is `NULL`, the callback is called with `buf`, which is the current rtp entry (this behaviour is briefly documented in the doc-comment on `do_in_path_and_pp`, although only `do_in_path` actually implements it) and if `path` is not `NULL` it is split on spaces and tabs at line 116, wildcard-expanded at line 126 and the callback is called on each expansion.

So `do_in_path_and_pp` does currently handle globbing and does also unconditionally split on spaces and tabs. `require()` currently works to find modules in packages (and still does on this PR branch) and currently breaks when the module name contains spaces or tabs (and still does on this PR branch).

I agree that these functions shouldn't be so fragile (I already ran into the problem that you still can't build Neovim from a path which contains whitespace, see #4621, but that depends on the build process for Nvim's dependencies too) but I'm not sure if that's in scope for this PR.

If it does get changed, the signature for many of these functions would have to change. `nvim_get_runtime_file` could, as a Lua function, take either a string or an array, but I'm not sure the current Lua API generation makes it easy to express this.

Alternatively, the string handling could be changed to handle escaped whitespace, but that might be even more effort than changing the signatures.

Reply:
we probably want a new `nvim_get_runtime_pattern` or something function later which takes a structural pattern (and is more efficient than spamming syscalls via string hacking), for now this PR is fine if it marginally improves things while still supporting all the existing cases. The exiting test for `fixtures/pack/foo/start/lua/bar.lua` needs to be complemented with  `fixtures/pack/foo/start/lua/spam/init.lua` to verify this.